dowhy/gcm/graph.py

```diff
@@ -190,3 +194,7 @@ def validate_node_has_causal_model(causal_graph: HasNodes, node: Any) -> None:
 def validate_node_in_graph(causal_graph: HasNodes, node: Any) -> None:
     if node not in causal_graph.nodes:
         raise ValueError("Node %s can not be found in the given graph!" % node)
+
+
+def _parent_samples_of(node: Any, scm, samples: pd.DataFrame) -> np.ndarray:
+    return samples[get_ordered_predecessors(scm.graph, node)].to_numpy()
```

Comment:
This function must go into a different module. The `graph` module is not the right place, because this function depends on scm (which you couldn't type here for that reason). Ways out:
- Duplicate this function in those modules where you need it (not a huge issue, because it's a oneliner, and it won't accumulate debt.
- introduce a separate helper module specifically for it. Note that such a module could host more functions like this. Because what we're looking at here, is a common pattern: topological traversal through the causal graph and separate treatment of root nodes and non-root nodes.

Reply:
Duplicated the code. The graph helper sounds like a good follow up. We can put things like ```is_root_node``` etc in there.